File: adapters/zapo/src/envelope.ts
/**
 * The wa-wire boundary format, written out.
 *
 * This is the same layout `wa-wire-contract` decodes, and the two must agree
 * byte for byte — a stanza encoded here is read there. The layout is fixed by
 * RFC-008:
 *
 * ```text
 * Envelope
 *   version      u16
 *   flags        u16     bit0 direction, bit1 frame_origin
 *   frame_len    u32
 *   frame        u8[frame_len]
 *   pt_count     u16
 *   pt_entries   PlaintextEntry[pt_count]
 *
 * PlaintextEntry
 *   path_len     u8
 *   path         u16[path_len]      little-endian child indices from the root
 *   status       u8
 *   payload_len  u32
 *   payload      u8[payload_len]
 * ```
 *
 * Little-endian throughout — unlike the stanza inside `frame`, which is
 * WhatsApp's own big-endian encoding and travels untouched.
 */

/** The contract version this module writes. */
export const CONTRACT_VERSION = 1

/** Bytes before the frame: version, flags, frame length. */
export const HEADER_LEN = 8

/** Which way a stanza was travelling. */
export enum Direction {
    Inbound = 0,
    Outbound = 1,
}

/** Whether the frame is the engine's own buffer or a re-encoding. */
export enum FrameOrigin {
    /** The buffer the engine's decoder consumed, verbatim. */
    Original = 0,
    /** Re-encoded from a decoded node, because the bytes were not reachable. */
    ReEncoded = 1,
}

/** Whether a plaintext entry holds usable bytes, and if not, why. */
export enum PlaintextStatus {
    Ok = 0,
    DecryptFailed = 1,
    Unsupported = 2,
}

/** One decrypted payload, addressed by the path of the node it came from. */
export interface Plaintext {
    /** Child indices from the root node. */
    readonly path: readonly number[]
    readonly status: PlaintextStatus
    /** Empty unless `status` is `Ok`. */
    readonly payload: Uint8Array
}

/** One stanza, ready to encode. */
export interface Stanza {
    readonly direction: Direction
    readonly frameOrigin: FrameOrigin
    /** The unpacked binary-node buffer. */
    readonly frame: Uint8Array
    readonly plaintexts?: readonly Plaintext[]
}

/** A value could not be represented in the envelope. */
export class EncodeError extends Error {
    public constructor(message: string) {
        super(message)
        this.name = 'EncodeError'
    }
}

const U8_MAX = 0xff
const U16_MAX = 0xffff
const U32_MAX = 0xffff_ffff

/** How many bytes `stanza` will encode to. */
export function encodedLength(stanza: Stanza): number {
    requireU32(stanza.frame.length, 'frame')
    const plaintexts = stanza.plaintexts ?? []
    requireU16(plaintexts.length, 'plaintext count')

    let total = HEADER_LEN + stanza.frame.length + 2
    for (const plaintext of plaintexts) {
        requireU8(plaintext.path.length, 'path')
        requireU32(plaintext.payload.length, 'payload')
        total += 1 + plaintext.path.length * 2 + 1 + 4 + plaintext.payload.length
    }
    return total
}

/**
 * Encode one stanza.
 *
 * Sized first and written once, so the buffer is never grown mid-write and the
 * length the decoder reads is the length that was produced.
 */
export function encodeEnvelope(stanza: Stanza): Uint8Array {
    const out = new Uint8Array(encodedLength(stanza))
    const view = new DataView(out.buffer)
    let at = 0

    view.setUint16(at, CONTRACT_VERSION, true)
    at += 2
    view.setUint16(at, packFlags(stanza), true)
    at += 2
    view.setUint32(at, stanza.frame.length, true)
    at += 4
    out.set(stanza.frame, at)
    at += stanza.frame.length

    const plaintexts = stanza.plaintexts ?? []
    view.setUint16(at, plaintexts.length, true)
    at += 2

    for (const plaintext of plaintexts) {
        out[at] = plaintext.path.length
        at += 1
        for (const component of plaintext.path) {
            requireU16(component, 'path component')
            view.setUint16(at, component, true)
            at += 2
        }
        out[at] = plaintext.status
        at += 1
        view.setUint32(at, plaintext.payload.length, true)
        at += 4
        out.set(plaintext.payload, at)
        at += plaintext.payload.length
    }

    return out
}

function packFlags(stanza: Stanza): number {
    let bits = 0
    if (stanza.direction === Direction.Outbound) bits |= 1 << 0
    if (stanza.frameOrigin === FrameOrigin.ReEncoded) bits |= 1 << 1
    return bits
}

/**
 * Check a length against the prefix that will carry it.
 *
 * Exported because the frame and payload limits need a 4 GiB buffer to reach
 * through `encodeEnvelope`, so the narrowing is checked here directly rather
 * than left as a branch no test can enter.
 */
export function fitsPrefix(value: number, bits: 8 | 16 | 32): boolean {
    const max = bits === 8 ? U8_MAX : bits === 16 ? U16_MAX : U32_MAX
    return Number.isInteger(value) && value >= 0 && value <= max
}

function requireU8(value: number, what: string): void {
    if (!fitsPrefix(value, 8)) {
        throw new EncodeError(`${what} of ${value} does not fit in a byte`)
    }
}

function requireU16(value: number, what: string): void {
    if (!fitsPrefix(value, 16)) {
        throw new EncodeError(`${what} of ${value} does not fit in 16 bits`)
    }
}

function requireU32(value: number, what: string): void {
    if (!fitsPrefix(value, 32)) {
        throw new EncodeError(`${what} of ${value} does not fit in 32 bits`)
    }
}

File: adapters/hypermeow/envelope.go
// Package wawire is the wa-wire adapter for the hypermeow engine.
//
// The boundary format, written out a third time. `wa-wire-contract` decodes
// it, the zapo adapter writes it in TypeScript, and this writes it in Go —
// because an adapter runs inside its engine, and this engine is Go. Rust in Go
// means cgo, and cgo in the per-stanza hot path is the cost the boundary exists
// to avoid.
//
// Three descriptions of one format that are only ever tested separately are
// three formats waiting to diverge, so the fixtures this package writes are
// read back by the Rust side and the ones it writes are read here.
//
// The layout is fixed by RFC-008:
//
//	Envelope
//	  version      u16
//	  flags        u16     bit0 direction, bit1 frame_origin
//	  frame_len    u32
//	  frame        u8[frame_len]
//	  pt_count     u16
//	  pt_entries   PlaintextEntry[pt_count]
//
//	PlaintextEntry
//	  path_len     u8
//	  path         u16[path_len]      little-endian child indices from the root
//	  status       u8
//	  payload_len  u32
//	  payload      u8[payload_len]
//
// Little-endian throughout — unlike the stanza inside frame, which is
// WhatsApp's own big-endian encoding and travels untouched.
package wawire

import (
	"encoding/binary"
	"errors"
	"fmt"
)

// ContractVersion is the boundary version this package writes.
//
// Not the WhatsApp protocol's version and never bumped for it: a protocol
// change crosses at L0 without the boundary noticing, which is what keeps a
// deployed adapter working when Meta ships something.
const ContractVersion uint16 = 1

// HeaderLen counts the bytes before the frame: version, flags, frame length.
const HeaderLen = 8

// MaxPathDepth is what the format can express: the length prefix is a u8.
//
// Not the same as what is reachable. `wa-wire-adapter` bounds a path at 64
// because the codec will not nest deeper than that, so a path past it cannot
// address anything — but that is the Rust SDK's rule and this is the format's,
// and a Go adapter has no business enforcing a limit the contract does not
// state.
const MaxPathDepth = 255

// Direction says which way a stanza was travelling.
type Direction uint8

const (
	// Inbound is a stanza received from the server.
	Inbound Direction = 0
	// Outbound is a stanza the client sent.
	Outbound Direction = 1
)

// FrameOrigin says whether the frame is the engine's own buffer.
type FrameOrigin uint8

const (
	// Original is the buffer the engine's decoder consumed, verbatim.
	Original FrameOrigin = 0
	// ReEncoded came from a decoded node, the bytes being unreachable.
	ReEncoded FrameOrigin = 1
)

// PlaintextStatus says whether an entry holds usable bytes, and if not, why.
type PlaintextStatus uint8

const (
	// StatusOk carries the plaintext.
	StatusOk PlaintextStatus = 0
	// StatusDecryptFailed means Signal refused it.
	StatusDecryptFailed PlaintextStatus = 1
	// StatusUnsupported means the adapter cannot decrypt this kind.
	StatusUnsupported PlaintextStatus = 2
	// StatusUnobserved means the node produced nothing the adapter saw.
	//
	// Claims less than the two failures above: an adapter watching plaintexts
	// appear can say a node produced none, but not why.
	StatusUnobserved PlaintextStatus = 3
)

// Plaintext is one decrypted payload, addressed by the path of the node it
// came from.
type Plaintext struct {
	// Path holds child indices from the root node.
	Path []uint16
	// Status says whether Payload is usable.
	Status PlaintextStatus
	// Payload is empty unless Status is StatusOk.
	Payload []byte
}

// Envelope is one stanza crossing the boundary.
type Envelope struct {
	Direction   Direction
	FrameOrigin FrameOrigin
	// Frame is the stanza exactly as the engine decoded it.
	Frame []byte
	// Plaintexts is the side table, one entry per node that decrypted.
	Plaintexts []Plaintext
}

// Errors an envelope can refuse to encode with. Each is a claim the format
// cannot carry, caught here rather than written out for a reader to trip over.
var (
	// ErrPathTooDeep is a path the length prefix cannot count.
	ErrPathTooDeep = errors.New("wawire: plaintext path is deeper than the contract can count")
	// ErrPayloadWithoutOk is a payload on an entry that did not decrypt.
	//
	// The contract says a non-Ok entry carries nothing, and a reader is
	// entitled to skip its bytes. Writing some would put data where nobody
	// looks.
	ErrPayloadWithoutOk = errors.New("wawire: only an Ok plaintext may carry a payload")
	// ErrTooManyPlaintexts is more entries than the count field can hold.
	ErrTooManyPlaintexts = errors.New("wawire: more plaintexts than the contract can count")
	// ErrFrameTooLong is a frame longer than the length field can hold.
	ErrFrameTooLong = errors.New("wawire: frame is longer than the contract can count")
)

// Encode writes the envelope in the boundary format.
//
// Allocates once and copies the frame, which an in-process consumer would not
// need — but this adapter's consumer is on the other side of a process or a
// language, so the copy is the crossing rather than an overhead on top of it.
func (e Envelope) Encode() ([]byte, error) {
	if len(e.Frame) > int(^uint32(0)) {
		return nil, ErrFrameTooLong
	}
	if len(e.Plaintexts) > int(^uint16(0)) {
		return nil, ErrTooManyPlaintexts
	}
	for index, plaintext := range e.Plaintexts {
		if len(plaintext.Path) > MaxPathDepth {
			return nil, fmt.Errorf("%w: entry %d has %d components", ErrPathTooDeep, index, len(plaintext.Path))
		}
		if plaintext.Status != StatusOk && len(plaintext.Payload) > 0 {
			return nil, fmt.Errorf("%w: entry %d", ErrPayloadWithoutOk, index)
		}
	}

	out := make([]byte, 0, e.encodedLen())
	out = binary.LittleEndian.AppendUint16(out, ContractVersion)
	out = binary.LittleEndian.AppendUint16(out, e.flags())
	out = binary.LittleEndian.AppendUint32(out, uint32(len(e.Frame)))
	out = append(out, e.Frame...)
	out = binary.LittleEndian.AppendUint16(out, uint16(len(e.Plaintexts)))
	for _, plaintext := range e.Plaintexts {
		out = append(out, uint8(len(plaintext.Path)))
		for _, component := range plaintext.Path {
			out = binary.LittleEndian.AppendUint16(out, component)
		}
		out = append(out, uint8(plaintext.Status))
		out = binary.LittleEndian.AppendUint32(out, uint32(len(plaintext.Payload)))
		out = append(out, plaintext.Payload...)
	}
	return out, nil
}

// flags packs direction and frame origin into the word the contract reads.
func (e Envelope) flags() uint16 {
	var flags uint16
	if e.Direction == Outbound {
		flags |= 1 << 0
	}
	if e.FrameOrigin == ReEncoded {
		flags |= 1 << 1
	}
	return flags
}

func (e Envelope) encodedLen() int {
	total := HeaderLen + len(e.Frame) + 2
	for _, plaintext := range e.Plaintexts {
		total += 1 + 2*len(plaintext.Path) + 1 + 4 + len(plaintext.Payload)
	}
	return total
}
